### PR TITLE
[REG-1105] feat: On-chain purchases

### PR DESCRIPTION
**Pull request by nickshatilo**
### 1. Contracts versioning
- [x] Make sure that the `patch` version of the contracts is increased if changes have been made to the `UNSRegistry`, `MintingManager`, `ProxyReader`, `ENSCustody` contracts.
- [ ] Make sure that the `minor` version of the contracts is increased if breaking changes have been made to the `UNSRegistry`, `MintingManager`, `ProxyReader`, `ENSCustody` contracts. It includes changes of interfaces.
### 2. Contracts licensing
- [ ] Make sure that no **SPDX-License-Identifier** defined in contracts.
- [x] Make sure that the **header** is added to the new contract files. 
  ```
  // @author Unstoppable Domains, Inc.
  // @date {Month} {Day}(ordinal), {Year}
  ```
### 3. Coverage
- [ ] Make sure that the coverage of contracts has not decreased and strive **100%**
### 4. Configs versioning
- [ ] Make sure that the version of `uns-config.json` is increased if changes have been made to the config.
- [ ] Make sure that the version of `ens-config.json` is increased if changes have been made to the config.
- [ ] Make sure that the version of `resolver-keys.json` is increased if changes have been made to the config.
### 5. Package versioning
- [ ] Make sure that the `patch` version of package is increased if valuable changes have been made to the package. It includes contracts update, configs update, etc.
- [ ] Make sure that the `major.minor` version of package is synced with version of `UNSRegistry` contract.
- [ ] Make sure that the `CHANGELOG` is updated with short description for the new version. 
### 6. Code review
- [ ] `resolver-keys.json` code review is required from **DevTools** team
- [ ] For all other changes code review is required from **Registry** team

### CODE REVIEW

File: contracts/MintingManager.sol [hunk 1]
```diff
@@ -196,6 +253,14 @@
         }
     }
 
+    function withdraw(address recepient) external onlyOwner {
+        payable(recepient).transfer(address(this).balance);
+    }
+
+    function withdraw(address token, address recepient) external onlyOwner {
+        IERC20Upgradeable(token).transfer(recepient, IERC20Upgradeable(token).balanceOf(address(this)));
+    }
```
[the remark this answers is not in the document]
Comment (nickshatilo) [reply contracts/MintingManager.sol]: Fixed

**Review by aquiladev — COMMENTED** [review]
do we want to try to test meta-txs?

File: contracts/MintingManager.sol [hunk 2]
```diff
@@ -196,6 +253,14 @@ contract MintingManager is ERC2771Context, MinterRole, Blocklist, Pausable, IMin
         }
     }
 
+    function withdraw(address recepient) external onlyOwner {
+        payable(recepient).transfer(address(this).balance);
```
[the remark this answers is not in the document]
Comment (nickshatilo) [reply contracts/MintingManager.sol:257]: Don't think we need event tracking / indexing for this type of operation, as long as it's done via multisig. Don't want to increase contract size unnecessary

[the remark this answers is not in the document]
Comment (aquiladev) [reply contracts/MintingManager.sol:257]: what if somebody will ask for report?

**Comment by nickshatilo** [comment]
@aquiladev Added a test for ERC20 meta tx purchase, but native tokens wouldn't work with meta txs.

File: contracts/MintingManager.sol [hunk 3]
```diff
@@ -13,16 +13,19 @@ import './roles/MinterRole.sol';
 import './utils/Blocklist.sol';
 import './utils/Pausable.sol';
 import './utils/Strings.sol';
+import '@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol';
+import '@openzeppelin/contracts-upgradeable/utils/cryptography/ECDSAUpgradeable.sol';
```
Comment (aquiladev) [inline contracts/MintingManager.sol:17]: pls move external import before internal

**Review by aquiladev — APPROVED** [review]

**Pull request merged**